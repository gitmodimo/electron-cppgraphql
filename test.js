describe("GraphQL native module tests", () => {
  const graphql = require(`./build/${
    process.env.USE_DEBUG_MODULE ? "Debug" : "Release"
  }/electron-cppgraphql`);

  it("starts the service", () => {
    graphql.startService();
  });

  let queryId = null;

  it("parses the introspection query", () => {
    queryId = graphql.parseQuery(`query {
        __schema {
            queryType {
                name
            }
            mutationType {
                name
            }
            subscriptionType {
                name
            }
            types {
                kind
                name
            }
        }
    }`);
    expect(queryId).not.toBeNull();
  });

  it("fetches introspection", async () => {
    expect(queryId).not.toBeNull();
    return expect(
      new Promise((resolve) => {
        let result = null;
        graphql.fetchQuery(
          queryId,
          "",
          "",
          (payload) => {
            result = JSON.parse(payload);
          },
          () => {
            resolve(result);
          }
        );
      })
    ).resolves.toMatchSnapshot();
  });

  it("cleans up after the query", () => {
    expect(queryId).not.toBeNull();
    graphql.unsubscribe(queryId);
    graphql.discardQuery(queryId);
    queryId = null;
  });

  let subscriptionId = null;

  it("parses subscription", () => {
    subscriptionId = graphql.parseQuery(`subscription {
        nodeChange(id: "ZmFrZVRhc2tJZA==") {
            id
            ...on Task {
                title
            }
        }
    }`);
    expect(subscriptionId).not.toBeNull();
  });

  let subscriptionPromise = null;
  let completedPromise = null;
  let completed = false;

  it("registers subscription", async () => {
    expect(subscriptionId).not.toBeNull();
    let resolveCompleted = null;
    completedPromise = new Promise((resolve) => {
      resolveCompleted = resolve;
    });
    subscriptionPromise = new Promise((resolve) => {
      graphql.fetchQuery(
        subscriptionId,
        "",
        "",
        (payload) => {
          resolve(JSON.parse(payload));
        },
        () => {
          completed = true;
          resolveCompleted();
        }
      );
    });
    expect(subscriptionPromise).not.toBeNull();
    expect(completed).toEqual(false);
  });

  let mutationId = null;

  it("parses mutation", () => {
    mutationId = graphql.parseQuery(`mutation {
        completedTask: completeTask(input: {id: "ZmFrZVRhc2tJZA==", isComplete: true, clientMutationId: "Hi There!"}) {
            completedTask: task {
                completedTaskId: id
                title
                isComplete
            }
            clientMutationId
        }
    }`);
    expect(mutationId).not.toBeNull();
  });

  it("applies mutation", async () => {
    expect(mutationId).not.toBeNull();
    return expect(
      new Promise((resolve) => {
        let result = null;
        graphql.fetchQuery(
          mutationId,
          "",
          "",
          (payload) => {
            result = JSON.parse(payload);
          },
          () => {
            resolve(result);
          }
        );
      })
    ).resolves.toMatchSnapshot();
  });

  it("cleans up after the mutation", () => {
    expect(mutationId).not.toBeNull();
    graphql.unsubscribe(mutationId);
    graphql.discardQuery(mutationId);
    mutationId = null;
  });

  it("updates subscriptions", () => {
    expect(subscriptionPromise).not.toBeNull();
    expect(subscriptionPromise).resolves.toMatchSnapshot();
  });

  it("cleans up after the subscription", async () => {
    expect(subscriptionId).not.toBeNull();
    expect(completed).toEqual(false);
    graphql.unsubscribe(subscriptionId);
    expect(completedPromise).not.toBeNull();
    await completedPromise;
    expect(completed).toEqual(true);
    graphql.discardQuery(subscriptionId);
    subscriptionId = null;
  });

  it("stops the service", () => {
    graphql.stopService();
  });
});
