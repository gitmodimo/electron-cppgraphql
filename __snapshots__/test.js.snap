// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL native module tests applies mutation 1`] = `
Object {
  "data": Object {
    "completedTask": Object {
      "clientMutationId": "Hi There!",
      "completedTask": Object {
        "completedTaskId": "ZmFrZVRhc2tJZA==",
        "isComplete": true,
        "title": "Don't forget",
      },
    },
  },
}
`;

exports[`GraphQL native module tests fetches introspection 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "mutationType": Object {
        "name": "Mutation",
      },
      "queryType": Object {
        "name": "Query",
      },
      "subscriptionType": Object {
        "name": "Subscription",
      },
      "types": Array [
        Object {
          "kind": "SCALAR",
          "name": "Boolean",
        },
        Object {
          "kind": "SCALAR",
          "name": "Float",
        },
        Object {
          "kind": "SCALAR",
          "name": "ID",
        },
        Object {
          "kind": "SCALAR",
          "name": "Int",
        },
        Object {
          "kind": "SCALAR",
          "name": "String",
        },
        Object {
          "kind": "ENUM",
          "name": "__TypeKind",
        },
        Object {
          "kind": "ENUM",
          "name": "__DirectiveLocation",
        },
        Object {
          "kind": "OBJECT",
          "name": "__Schema",
        },
        Object {
          "kind": "OBJECT",
          "name": "__Type",
        },
        Object {
          "kind": "OBJECT",
          "name": "__Field",
        },
        Object {
          "kind": "OBJECT",
          "name": "__InputValue",
        },
        Object {
          "kind": "OBJECT",
          "name": "__EnumValue",
        },
        Object {
          "kind": "OBJECT",
          "name": "__Directive",
        },
        Object {
          "kind": "SCALAR",
          "name": "ItemCursor",
        },
        Object {
          "kind": "SCALAR",
          "name": "DateTime",
        },
        Object {
          "kind": "ENUM",
          "name": "TaskState",
        },
        Object {
          "kind": "INPUT_OBJECT",
          "name": "CompleteTaskInput",
        },
        Object {
          "kind": "INPUT_OBJECT",
          "name": "ThirdNestedInput",
        },
        Object {
          "kind": "INPUT_OBJECT",
          "name": "FourthNestedInput",
        },
        Object {
          "kind": "INPUT_OBJECT",
          "name": "SecondNestedInput",
        },
        Object {
          "kind": "INPUT_OBJECT",
          "name": "FirstNestedInput",
        },
        Object {
          "kind": "INTERFACE",
          "name": "Node",
        },
        Object {
          "kind": "UNION",
          "name": "UnionType",
        },
        Object {
          "kind": "OBJECT",
          "name": "Query",
        },
        Object {
          "kind": "OBJECT",
          "name": "PageInfo",
        },
        Object {
          "kind": "OBJECT",
          "name": "AppointmentEdge",
        },
        Object {
          "kind": "OBJECT",
          "name": "AppointmentConnection",
        },
        Object {
          "kind": "OBJECT",
          "name": "TaskEdge",
        },
        Object {
          "kind": "OBJECT",
          "name": "TaskConnection",
        },
        Object {
          "kind": "OBJECT",
          "name": "FolderEdge",
        },
        Object {
          "kind": "OBJECT",
          "name": "FolderConnection",
        },
        Object {
          "kind": "OBJECT",
          "name": "CompleteTaskPayload",
        },
        Object {
          "kind": "OBJECT",
          "name": "Mutation",
        },
        Object {
          "kind": "OBJECT",
          "name": "Subscription",
        },
        Object {
          "kind": "OBJECT",
          "name": "Appointment",
        },
        Object {
          "kind": "OBJECT",
          "name": "Task",
        },
        Object {
          "kind": "OBJECT",
          "name": "Folder",
        },
        Object {
          "kind": "OBJECT",
          "name": "NestedType",
        },
        Object {
          "kind": "OBJECT",
          "name": "Expensive",
        },
      ],
    },
  },
}
`;

exports[`GraphQL native module tests updates subscriptions 1`] = `
Object {
  "data": Object {
    "nodeChange": Object {
      "id": "ZmFrZVRhc2tJZA==",
      "title": "Don't forget",
    },
  },
}
`;
